src/Bicep.Core/Syntax/AstVisitor.cs

```diff
@@ -0,0 +1,349 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Bicep.Core.Parsing;
+
+namespace Bicep.Core.Syntax
+{
+    /// <summary>
+    /// Visits an <see href="https://en.wikipedia.org/wiki/Abstract_syntax_tree">abstract syntax tree (AST)</see>.
+    /// </summary>
+    public abstract class AstVisitor : ISyntaxVisitor
```

Comment:
It might save some repetition if `AstVisitor` were a specialization of rather than substitute for the existing `SyntaxVisitor`. `AstVisitor` could extend `SyntaxVisitor` and override the CST-only node visitation methods (for literals, trivia, and tokens) with no-ops.

Reply:
Good point. I'll create an `SyntaxVisitor` for sharing common methods like `Visit` and `VisitInternal`. I would like to avoid having `AstVisitor` be a child class of `CstVisitor` though. If a new syntax node type is added in the future, the current hierarchy (ISyntaxVisitor -> AstVisitor, ISyntaxVisitor -> CstVisitor) forces people to add Visit* method for the new syntax node type for both visitors.

Reply:
Do you need a `CstVisitor`? The only difference in behavior between the two visitor types is whether CST-specific node visitation methods are no-ops.

Reply:
That's not the only difference. Many of the `AstVisitor` methods are also customized to avoid calls to the no-op visitation methods. For example:

```C#
// AstVisitor
public virtual void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
{
    this.VisitNodes(syntax.LeadingNodes);
    this.Visit(syntax.Name);
    this.Visit(syntax.Type);
    this.Visit(syntax.Value);
}

// CstVisitor
public virtual void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
{
    this.VisitNodes(syntax.LeadingNodes);
    this.Visit(syntax.Keyword);
    this.Visit(syntax.Name);
    this.Visit(syntax.Type);
    this.Visit(syntax.ExistingKeyword);
    this.Visit(syntax.Assignment);
    this.Visit(syntax.Value);
}
```